pynitrokey/test_secrets_app.py

```diff
@@ -1337,3 +1375,166 @@ def test_password_safe_just_pws_entry(secretsAppResetLogin):
+            data_raw=helper_get_padded(b"1"),
+        )
+
+    # registration on the special-named slots works
+    # TODO Do not offer actual names in CLI, but rather select slot by number
+    for i, slot in enumerate([b"HmacSlot2", b"HmacSlot1"]):
+        secretsAppRaw.register(
+            slot,
+            secret=i.to_bytes(1, "little") * 20,
+            kind=Kind.Hmac,
+            pin_based_encryption=False,
+        )
+
+    for slot in [YK_P1_CMD_HMAC_2, YK_P1_CMD_HMAC_1]:
+        # calculation on the hmac slot works
+        status, data = helper_send_receive_ins(
+            secretsAppRaw,
+            YK_API_REQ,
+            p1=slot,
+            le=20,
+            data_raw=helper_get_padded(slot.to_bytes(1, "little") * 63),
+        )
+        assert len(data) == 20
+
+        # different input gives different output
+        status, data = helper_send_receive_ins(
+            secretsAppRaw,
+            YK_API_REQ,
+            p1=slot,
+            le=20,
+            data_raw=helper_get_padded(b"1" * 63),
+        )
+        status, data2 = helper_send_receive_ins(
+            secretsAppRaw,
+            YK_API_REQ,
+            p1=slot,
+            le=20,
+            data_raw=helper_get_padded(b"2" * 63),
+        )
+        assert data != data2
+
+        # same input gives same output
+        status, data3 = helper_send_receive_ins(
+            secretsAppRaw,
+            YK_API_REQ,
+            p1=slot,
+            le=20,
+            data_raw=helper_get_padded(b"2" * 63),
+        )
+        assert data3 == data2
+
+    # As the last step, set the secret as the one used in the KeepassXC tests
+    # $ ninja testykchallengeresponsekey && ./tests/testykchallengeresponsekey
+    secret = binascii.a2b_hex(
+        "1c e3 0f d7 8d 20 dc fa 40 b5 0c 18 77 9a fb 0f 02 28 8d b7".replace(" ", "")
+    )
+    for slot_name in [b"HmacSlot2", b"HmacSlot1"]:
+        secretsAppRaw.register(
+            slot_name,
+            secret=secret,
+            kind=Kind.Hmac,
+        )
+
+    # Do not allow to register secret with different lengths than expected 20 bytes
+    for secret_len in [18, 21, 200]:
+        with pytest.raises(SecretsAppException, match="IncorrectDataParameter"):
+            secretsAppRaw.register(
+                b"HmacSlot2",
+                secret=b"x" * secret_len,
+                kind=Kind.Hmac,
+            )
+
+    # Test various challenge lengths against local calculations
+
+    """
+    Comment from the KeepassXC implementation, copied verbatim:
+    /*
+     * The challenge sent to the Yubikey should always be 64 bytes for
+     * compatibility with all configurations.  Follow PKCS7 padding.
+     *
+     * There is some question whether or not 64 bytes fixed length
+     * configurations even work, some docs say avoid it.
+     *
+     * In fact, the Yubikey always assumes the last byte (nr. 64)
+     * and all bytes of the same value preceding it to be padding.
```

Comment:
Do you have a link to the documentation of this padding?

Reply:
I wish! I found this one in the KeepassXC source code :(
Have not searched further once this worked.

Reply:
Where is it in the KeepassXC source code?

Reply:
https://github.dev/Nitrokey/keepassxc/blob/cf819e0a3f5664fb0e1705217dbebbdf704bdc34/src/keys/drivers/YubiKeyInterfacePCSC.cpp

Reply:
Thanks.

So that means that the padding is naturally ambiguous? Do we need to be compatible with Yubikey on this?

Reply:
I think it has to be bug-for-bug compatibility, if we plan to support other applications using it, not only KeepassXC.
Doing this right will not be Yubikey compatible, which misses the point in my opinion.

Reply:
In the end it does not matter in this case, as payload bigger than 63 bytes can't be used anyway.